### PR TITLE
API: add more oidc configuration settings

### PR DESCRIPTION
**Pull request by zhaohuabing**
This RP adds more configuration settings to the OIDC API.

Related issue: https://github.com/envoyproxy/gateway/issues/3405 https://github.com/envoyproxy/gateway/issues/3395 https://github.com/envoyproxy/gateway/issues/3462

@coro @denniskniep

### CODE REVIEW

File: api/v1alpha1/oidc_types.go [hunk 1]
```diff
@@ -50,6 +51,35 @@ type OIDC struct {
 	// The path to log a user out, clearing their credential cookies.
 	// If not specified, uses a default logout path "/logout"
 	LogoutPath *string `json:"logoutPath,omitempty"`
+
+	// ForwardBearerToken indicates whether the Envoy should forward the access
+	// token as a bearer token in the "Authorization" header to the backend.
+	// If not specified, defaults to false.
+	// +optional
+	ForwardBearerToken *bool `json:"forwardBearerToken,omitempty"`
```
Comment (zhaohuabing) [inline api/v1alpha1/oidc_types.go:59]: This changes the current default behavior, which forwards the access token. 

We should mention this is the release doc.

Comment (arkodg) [reply api/v1alpha1/oidc_types.go:62]: is there a batter name for `ForwardAccessTokenAsBearerToken` ?

File: api/v1alpha1/oidc_types.go [hunk 2]
```diff
@@ -48,8 +49,50 @@ type OIDC struct {
...
+	// of the authorization server if it is provided. This field is only used when
+	// the expiry time is not provided by the authorization.
+	//
+	// If not specified, defaults to 0. In this case, the "expires_in" field in
+	// the authorization response must be set by the authorization server, or the
+	// OAuth flow will fail.
+	//
+	// +optional
+	// +notImplementedHide
+	DefaultTokenExpireTime *metav1.Duration `json:"defaultTokenExpireTime,omitempty"`
```
Comment (arkodg) [inline api/v1alpha1/oidc_types.go:75]: should this be `DefaultIdTokenExpireTime` ?

File: api/v1alpha1/oidc_types.go (hunk 1 above)
Comment (zhaohuabing) [reply api/v1alpha1/oidc_types.go:62]: > is there a batter name for `ForwardAccessTokenAsBearerToken` ?

I think Bearer Token is normally an Access Token, so we probably don't need to emphasize that in the name?

File: api/v1alpha1/oidc_types.go (hunk 2 above)
Comment (zhaohuabing) [reply api/v1alpha1/oidc_types.go:75]: The Envoy doc is somewhat vague in[ its description](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/oauth2/v3/oauth.proto#envoy-v3-api-msg-extensions-filters-http-oauth2-v3-oauth2config) of `DefaultTokenExpireTime`.

However, after digging into the code, it's clear that this applies to both the ID token and Access Token.

https://github.com/envoyproxy/envoy/blob/a112766683a6b82d20843c70b8aa4d7eda29048e/source/extensions/filters/http/oauth2/oauth_client.cc#L183

File: api/v1alpha1/oidc_types.go (hunk 1 above)
Comment (zhaohuabing) [reply api/v1alpha1/oidc_types.go:62]: Thought about it again, it's better to make it clear even though the name becomes a bit longer.

File: api/v1alpha1/oidc_types.go (hunk 2 above)
Comment (zhaohuabing) [reply api/v1alpha1/oidc_types.go:75]: Or `DefaultIdAndAccessTokenExpireTime`, that's quite lengthy :-)

File: api/v1alpha1/oidc_types.go (hunk 1 above)
Comment (zhaohuabing) [reply api/v1alpha1/oidc_types.go]: Change it to ForwardAccessToken and add a comment to explain that the access token will be forwarded via the Authorization header Bearer scheme to the upstream.

**Review by zetaab — APPROVED** [review]
LGTM, going to need all of these :)

File: api/v1alpha1/oidc_types.go [hunk 3]
```diff
@@ -48,8 +49,49 @@ type OIDC struct {
...
+	// of the authorization server if it is provided. This field is only used when
+	// the expiry time is not provided by the authorization.
+	//
+	// If not specified, defaults to 0. In this case, the "expires_in" field in
+	// the authorization response must be set by the authorization server, or the
+	// OAuth flow will fail.
+	//
+	// +optional
+	// +notImplementedHide
+	DefaultTokenExpireIn *metav1.Duration `json:"defaultTokenExpireIn,omitempty"`
```
Comment (arkodg) [inline api/v1alpha1/oidc_types.go:74]:
```suggestion
	DefaultTokenExpiresIn *metav1.Duration `json:"defaultTokenExpireIn,omitempty"`
```

Comment (arkodg) [reply api/v1alpha1/oidc_types.go:74]: we need a better name here `DefaultTokenExpiresIn` that encapsulates default expiry duration for the token

Comment (zhaohuabing) [reply api/v1alpha1/oidc_types.go:74]: `DefaultTokenExpiryDuration`
`DefaultTokenTTL`

Comment (zirain) [reply api/v1alpha1/oidc_types.go:74]: +1 on `DefaultTokenTTL`

**Review by arkodg — APPROVED** [review]

**Pull request merged**